runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.nonrepudiation.api.v1
+
+import java.io.ByteArrayInputStream
+import java.security.cert.{CertificateFactory, X509Certificate}
+
+import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.server.{Route, StandardRoute, ValidationRejection}
+import com.daml.nonrepudiation.{CertificateRepository, FingerprintBytes}
+import com.google.common.io.BaseEncoding
+import org.slf4j.{Logger, LoggerFactory}
+import spray.json.{DefaultJsonProtocol, RootJsonFormat}
+
+import scala.util.Try
+
+private[api] final class CertificatesEndpoint private (certificates: CertificateRepository)
+    extends CertificatesEndpoint.JsonProtocol {
+
+  import CertificatesEndpoint._
+
+  private def putCertificate(certificate: X509Certificate): Route =
+    handleExceptions(logAndReport(logger)(UnableToAddTheCertificate)) {
+      val fingerprint = certificates.put(certificate)
+      complete(encode(fingerprint))
+    }
+
+  private def getCertificate(fingerprint: FingerprintBytes): Route =
+    handleExceptions(logAndReport(logger)(UnableToRetrieveTheCertificate)) {
+      certificates.get(fingerprint).fold(reject)(certificate => complete(encode(certificate)))
+    }
+
+  private val route: Route =
+    concat(
+      put {
```

Comment:
Related question to the one above, if we’re not idempotent, we probably want post. Couldn’t find a test to check whether we are.

Reply:
Follow-up in https://github.com/digital-asset/daml/pull/8982#discussion_r585666749